CONTRIBUTING.md

```diff
@@ -0,0 +1,105 @@
+
+To clone the project's sources and install its various dependencies, run:
+
+```sh
+git clone https://github.com/mirage/alcotest.git    # Get the repository
+cd alcotest
+
+opam switch create ./ ocaml-base-compiler.4.10.0    # OPTIONAL: install a project-local Opam switch
+opam install -t --deps-only .                       # Install regular and test dependencies
+```
+
+Now you can work on the project using the standard Dune commands:
+
+```sh
+dune build    # Build all libraries, executables and tests
+dune test     # Run the test-suite
+dune clean    # Delete all artefacts
+```
+
+## Project structure
+
+The high-level project structure is as follows:
+
+```sh
+├── src/
+│   ├── alcotest-engine/
+│   │   ├── core.ml           # The main test runner, parameterised over...
+│   │   ├── monad.ml          # ... a concurrency monad
+│   │   ├── platform.ml       # ... a platform implementation / OS bindings
+│   │   │
+│   │   ├── cli.ml            # Extends `core.ml` with a command-line API
+│   │   │
+│   │   ├── pp.ml             # Pretty-printers for Alcotest output
+│   │   ├── test.ml           # Combinators for test assertions (with `Alcotest.check`)
+│   │   └── utils.ml          # Standard-library extension
+│   │
+│   │                         # Specific backends:
+│   ├── alcotest/             # - Unix-specific API
+│   ├── alcotest-async/       # - Extends `alcotest/` for test-suites using Async concurrency
+│   ├── alcotest-lwt/         # - Extends `alcotest/` for test-suites using Lwt concurrency
+│   └── alcotest-mirage/      # - MirageOS-specific API
+│
+├── test/                     # Project tests (see '§ Expect tests' below)
+│
+├── dune-project              # Project metadata (opam files are generated from
+├── alcotest-async.opam       # data in the `dune-project` file by running
+├── alcotest-lwt.opam         # `dune build @install`).
+├── alcotest-mirage.opam
+├── alcotest.opam
+│
+├── alcotest-help.txt         # Manpage output (generated by `dune test --auto-promote`)
+├── dune
+│
+├── CHANGES.md
+└── README.md
+```
+
+## Expect tests
+
+The repository contains a number of 'expect' tests in
+https://github.com/mirage/alcotest/tree/master/test/e2e. These are short,
+self-contained usages of the Alcotest API, with the corresponding output
+snapshotted in an `.expected` file:
+
+```sh
+test/e2e/alcotest/passing/
+├── basic.ml          # The test case
+├── basic.expected    # The expected output of the test case
+└── basic.opts        # OPTIONAL: command-line options to pass to the test
+```
+
+If you change the Alcotest output in some way, you will likely need to update
+the expect test snapshots:
+
+1. run `dune test` to get a list of diffs to the tests,
+1. check that the diffs are as expected,
+1. run `dune promote` to update the snapshots in the repository.
+
+To add a new test, create a new `foo.{ml,expected}` pair in one of the test
+directories and run `dune test --auto-promote` to automatically generate the
```

Comment:
How does one use the quick/slow test feature with dune? 

I tried it with both `dune test` and `dune runtest` but couldn't figure it out.

Reply:
It's possible to do it with `ALCOTEST_QUICK_TESTS=true dune test`, but that's not particularly ergonomic :wink:. We probably shouldn't have this hard-coded into our API, since it seems like a fairly special-case feature to me (c.f. https://github.com/mirage/alcotest/issues/124).

Reply:
Ah cool! Does this mean there isn't an ergonomic solution for filtering tests via `dune (run)test`?

Reply:
There is not.

I generally do `dune exec test/foo/test.exe -- test <filter>` when I want to do this, which also works to pass CLI options but only runs a single test binary at a time.